compiler/rustc_codegen_llvm/src/intrinsic.rs

```diff
@@ -1704,6 +1704,93 @@ unsupported {} from `{}` with element `{}` of size `{}` to `{}`"#,
     bitwise_red!(simd_reduce_all: vector_reduce_and, true);
     bitwise_red!(simd_reduce_any: vector_reduce_or, true);
 
+    if name == sym::simd_cast_ptr {
+        require_simd!(ret_ty, "return");
+        let (out_len, out_elem) = ret_ty.simd_size_and_type(bx.tcx());
+        require!(
+            in_len == out_len,
+            "expected return type with length {} (same as input type `{}`), \
+                  found `{}` with length {}",
+            in_len,
+            in_ty,
+            ret_ty,
+            out_len
+        );
+
+        match in_elem.kind() {
+            ty::RawPtr(p) => require!(
+                p.ty.is_sized(bx.tcx.at(span), ty::ParamEnv::reveal_all()),
```

Comment:
imho this should check for [thin](https://github.com/rust-lang/rust/issues/81513) pointers rather than pointers to sized types, we should be able to have vectors of pointers to extern types. Afaict you need to call [`ptr_metadata_ty`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.ptr_metadata_ty) and handle the results. Maybe ask on Zulip if there's a handy `is_thin_ptr` function or something...

Reply:
I believe extern types are represented by `Opaque`, and this line is hinting that they must be sized, anyway:
https://github.com/rust-lang/rust/blob/c07a8b4e09f356c7468b69c50cac7fc5b5000b8a/compiler/rustc_middle/src/ty/sty.rs#L2123

Reply:
extern types are `ty::Foreign` and unsized, with metadata `()`:
https://github.com/rust-lang/rust/blob/5b8cf49c51833ee5d27ae2e8e179337dbb9f14d7/compiler/rustc_type_ir/src/sty.rs#L52-L53

imho you should not check specifically for `ty::Foreign` and instead use `ptr_metadata_ty`:
```rust
// probably right, should be double-checked though
let (metadata_ty, need_sized_check) = pointee_ty.ptr_metadata_ty(...);
if need_sized_check {
    // idk how to do this
    require_sized_trait(pointee_ty);
}
// now you can use metadata_ty
if !is_unit_type(metadata_ty) {
    // error
}
```
something like that...though maybe a better route is to get the in-memory layout of the pointer type and check that it's just a single pointer.

Reply:
I'm having a bit of trouble figuring out what "normalize" in this context means.  I only see one use of this function anywhere in rustc...

Reply:
I asked on Zulip: https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/help.20checking.20for.20thin.20pointers.20in.20codegen/near/295534523

Reply:
I believe it should be good, now